doc/design/programming_solver.md

```diff
@@ -0,0 +1,141 @@
+- x,y >= 0 (sign restriction)
+
+
+To solve this problem, we can write objective and constraints using SQL, then SQLFlow should call the corresponding "solver" (e.g. [pyomo](http://www.pyomo.org/) and [GLPK](https://www.gnu.org/software/glpk/)) to find the result and output the result into the result table.
+
+## Grammar Design
+
+Generally, SQLFlow users can write one SQL statement to describe the programming problem:
+
+```sql
+SELECT * FROM train_table
+TO SOLVE LP
+WITH
+     objective="sum([(train_table.price[i] - train_table.materials_cost[i] - train_table.other_cost[i]) * @X[i] for i in @X])",
+     constraints=["sum([train_table.finishing[i] * @X[i] for i in @X]) <= 100",
+                  "sum([train_table.carpentry[i] * @X[i] for i in @X]) <= 80",
+                  "@X[i] <= train_table.max_num[i] for i in @X",
+                  "@X[i] >= 0 for i in @X"],
+     var_name_col="type",
+     var_type="Integers",
+     solver="glpk"
+INTO result_table;
+```
+
+The `train_table` looks like:
+
+|  type   | price | materials_cost | other_cost | finishing | carpentry | max_num |
+| ------- | ----- | -------------- | ---------- | --------- | --------- | ------- |
+| soldier | 27    | 10             | 14         | 2         | 1         | 40      |
+| train   | 21    | 9              | 10         | 1         | 1         | 10000   |
+
+Note that we create this table to store the variables to form the objective and constraints so that when we have hundreds of variable types (e.g. the company actually manufactures 1000 products), the SQL statement will keep the same.
+
+In the SQL statement:
+
+- `TO SOLVE LP`: set to use the "Linear Programming Solver". The notation `LP` means "Linear Programming Solver", and we can have other solvers like `QP` for "Quadratic programming Solver" etc.
+- `INTO result_table`: set the result table name.
+- `WITH` attributes:
+    - objective: **required**, an expression string that describes the objective. Notation `@X` will be replaced to input dataframe when generating Python code.
+    - constraints: **required**, a list of expression strings that describe the constraints.
+    - var_name_col: **required**, specify one column that stores the variable name.
+    - var_type: **required**, specify the variable type, can be `Integers`, `NonNegativeIntegers`, `Reals` etc.
+    - solver: *optional*, solver tool to use, default: glpk.
+
+After the SQL statement finishes execution, the result table `result_table` should look like:
+
+| type    | result |
+| ------  | ------ |
+| soldier | 20     |
+| train   | 60     |
+
+## Implementation
+
+1. Update current extended syntax parser to support `TO SOLVE` clause.
+1. Add IR struct to represent `TO SOLVE` clause.
+1. Create a table to store the result.
+1. Add code generator to generate code like below example to run.
+1. The generated code should be able to output the result to the result table.
+
+
+## Example of Generated Code
```

Comment:
We'd better separate syntax in sqlflow from pyomo syntax, as we can convert the syntax in sqlflow to other modeling language more than pyomo in the future. It will help us to extend to other solvers.

Reply:
Agreed. @samuel5200.  I have the same concern in https://github.com/sql-machine-learning/sqlflow/pull/2333#discussion_r429723331